### ISSUE TITLE
EC2 instance type discovery

### ISSUE THREAD

**Issue by ewbankkit** [terraform-providers/terraform-provider-aws#10989]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Amazon EC2 makes it easier for customers to discover and compare EC2 instance types.
Customers can programmatically access instance type specifications using new APIs: `DescribeInstanceType`s and `DescribeInstanceTypeOfferings`.

### New or Affected Resource(s)

* `aws_instance_types`
* `aws_instance_type`

### Potential Terraform Configuration

```hcl
data "aws_instance_types" "nitro" {
  filter {
    name   = "hypervisor"
    values = ["nitro"]
  }
}
```

### References

[Announcement](https://aws.amazon.com/about-aws/whats-new/2019/11/amazon-ec2-makes-it-easier-for-customers-to-discover-and-compare-ec2-instance-types/).
[User guide](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-discovery.html).
API references:
* [`DescribeInstanceTypes`](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeInstanceTypes.html)
* [`DescribeInstanceTypeOfferings`](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeInstanceTypeOfferings.html)

Requires [AWS SDK v1.25.41](https://github.com/aws/aws-sdk-go/releases/tag/v1.25.41):
* https://github.com/terraform-providers/terraform-provider-aws/pull/10955

**Comment by bflad** [comment]
Submitted new data sources for EC2 Instance Type Offerings: https://github.com/terraform-providers/terraform-provider-aws/pull/12139

**Comment by bflad** [comment]
Two new data sources:
* New Data Source: `aws_ec2_instance_type_offering`
* New Data Source: `aws_ec2_instance_type_offerings`

Have been merged and will release with version 2.52.0 of the Terraform AWS Provider, Thursday this week. Leaving this open for `aws_ec2_instance_type`/`aws_ec2_instance_types` implementations.

**Comment by helgi** [comment]
I noticed in #12139 that the return values are mostly focused on `instance_types` and does not provide the `location` and `location_type` that https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-instance-type-offerings.html offers. Was there a particular reason?

It is useful seeing what AZs are available for a given instance type within a region:

M4.xlarge in us-east-1
`aws --region us-east-1 ec2 describe-instance-type-offerings --filters "Name=instance-type,Values=m4.xlarge" --location-type availability-zone` returns us-east-1a, us-east-1b, us-east-1c, us-east-1d, and us-east-1e

where as M5.xlarge
`aws --region us-east-1 ec2 describe-instance-type-offerings --filters "Name=instance-type,Values=m5.xlarge" --location-type availability-zone` correctly excludes us-east-1e where it is not available.

One could imagine, through some hop jumping, you can get all AZs in a region, fetch the AZs for your instance type and then find the delta and pass into the blacklist for `aws_availability_zones` (with any additional filtering you need).
